res/controllers/Pioneer-DDJ-SR2-scripts.js

```diff
@@ -0,0 +1,1317 @@
+        midi: [0x90 + channelOffset, 0x10],
+        key: "loop_in",
+    });
+
+    this.loopOut = new components.Button({
+        midi: [0x90 + channelOffset, 0x11],
+        key: "loop_out",
+    });
+
+    this.slotSelect = new components.Button({
+        midi: [0x90 + channelOffset, 0x4C],
+        key: "quantize",
+        type: components.Button.prototype.types.toggle,
+    });
+
+    this.reloopExit = new components.Button({
+        midi: [0x90 + channelOffset, 0x4D],
+        key: "reloop_toggle",
+    });
+
+    this.censor = new components.Button({
+        midi: [0x90 + channelOffset, 0x15],
+        key: "reverseroll"
+    });
+
+    this.reverse = new components.Button({
+        midi: [0x90 + channelOffset, 0x38],
+        key: "reverse",
+        type: components.Button.prototype.types.toggle
+    });
+
+    this.slipButton = new components.Button({
+        midi: [0x90 + channelOffset, 0x40],
+        key: "slip_enabled",
+        type: components.Button.prototype.types.toggle
+    });
+
+    this.vinylButton = new components.Button({
+        midi: [0x90 + channelOffset, 0x17],
+        type: components.Button.prototype.types.toggle,
+        trigger: function() {
+            this.output(theDeck.jog.vinylMode);
+        },
+        input: function(channel, control, value, status, group) {
+            if (this.isPress(channel, control, value, status, group)) {
+                theDeck.jog.vinylMode = !theDeck.jog.vinylMode;
+                this.send(theDeck.jog.vinylMode);
+            }
+        }
+    });
+    this.vinylButton.trigger();
+
+    this.brakeStopButton = new components.Button({
+        midi: [0x90 + channelOffset, 0x79],
+        type: components.Button.prototype.types.toggle,
+        inToggle: function() {
+            // slow down the track
+            engine.brake(deckNumber, true);
+        }
+    });
+    this.softStartButton = new components.Button({
+        midi: [0x90 + channelOffset, 0x0A],
+        type: components.Button.prototype.types.toggle,
+        inToggle: function() {
+            engine.softStart(deckNumber, true);
+        }
+    });
+
+    this.gridSetButton = new components.Button({
+        midi: [0x90 + channelOffset, 0x64],
+        inKey: "beats_translate_curpos"
+    });
+
+    this.keySync = new components.Button({
+        midi: [0x90 + channelOffset, 0x70],
+        key: "sync_key"
+    });
+
+    this.keyShiftDown = new components.Button({
+        midi: [0x90 + channelOffset, 0x72],
+        key: "pitch_down"
+    });
+
+    this.keyShiftUp = new components.Button({
+        midi: [0x90 + channelOffset, 0x73],
+        key: "pitch_up"
+    });
+
+    this.keyReset = new components.Button({
+        midi: [0x90 + channelOffset, 0x71],
+        inKey: "reset_key"
+    });
+    this.wheelRing = new WheelRing({
+        midi: [0xBB, 0x20 + channelOffset],
+        color: DDJSR2.wheelColor[deckNumber],
+        deckNumber: deckNumber,
+    });
```

Comment:
```suggestion
    switch (DDJSR2.wheelColor[deckNumber]) {
    case "TRACK":
        this.wheelRing = new WheelRing({
            midi: [0xBB, 0x20 + channelOffset],
            group: this.group,
            outKey: "track_color",
            output: function(colorRGB) { this.sendIfChanged(this.colorMapper.getValueForNearestColor(colorRGB)); };
        });
        break;
    case "POSITION":
        this.wheelRing = new WheelRing({
            midi: [0xBB, 0x20 + channelOffset],
            group: this.group,
            outKey: "position",
            revolutionsPerSecond: DDJSR2.scratchSettings.vinylSpeed / 60;
            output: function(position) {
                // Every time the playposition  changes, update the wheel color.
                // Timing calculation is handled in seconds!!
                const elapsedTime = position * engine.getValue(this.group, "duration");;
                // Determine the current position in the color cycle
                // The modulo operation ensures the color value stays within the available range
                // The multiplication by maxColorValue ensures that the position is spread over the entire range of indexed colors.
                this.send(Math.round((this.revolutionsPerSecond * elapsedTime * maxColorValue) % maxColorValue));
            }
        });
        break;
    default:
        this.wheelRing = new WheelRing({
            midi: [0xBB, 0x20 + channelOffset],
            group: this.group,
            outValueScale: () => DDJSR2.wheelColor[deckNumber]
        });
    };
```

Reply:
Pushed a new update to the wheelRing code, to leverage the 500ms timer, and ensure that the endWarning occurs even in TRACK mode. Also fixes the jog wheel color on load of a new track.

Reply:
can you elaborate on why you chose what you pushed over what I proposed? I think what you pushed is much less clean...

Reply:
because it didn't work. kept getting messages about .call'ing on undefined, for components.Component.output.call and send.call, couldn't work out why it wasn't working. Knew my code worked, just needed the TRACK mode working, and it synced with 500ms, you provided the glue for that.

Reply:
> components.Component.output.call and send.call,

yeah, should have been `components.Component.prototype.output.call`. my bad.

Reply:
If it did work, still needed the endWarning blink code in it, going from red to white alternating with each 500ms, which is what this.blink = !this.blink does. .fixcolor is there, as I found a bug, Once endWarning happens, if one changes track, and it happens to be the same colored track, as the engine thinks there is no track color change, it doesn't run the output, so the new track doesn't reset the jog wheel color. using track_loaded is forcing the refresh.

Reply:
The endWarning code was effectively already integrated (for every mode). It just alternated between the wheel being off and the color produced by the mode instead. You can change the color sent during the off-cycle easily. I didn't know you wanted red-white-alternating as the warning pattern.

Reply:
primarily that the concerns are more cleanly separated. The stock wheelring is the component that implements the end_of_track warning logic. If you were to add another mode, you can just create another customized instance of the wheelring and you get the end_of_track logic for free. See the open-closed principle for more. In addition there are also a couple little things:
1. indicator_500ms is used wrong. You should respond to the actual value of the CO (1 is on, 0 is off) instead of tracking the state yourself (the `.blink` property you introduced). Otherwise depending on the exact timing, you may get a wheelrings that blink in phase and other times 180 degrees out-of-phase. 
2. returning some sentinel value `"EOT"` in getPositionColor makes it clear that your code is not separated into clean chunks (see "separation of concerns").
3. the need for `fix_color` is yet another workaround that indicates the your solution is suboptimal. in the solution I posted it should automatically use the correct color (at least in principle; there may be still be bugs that occur when coding blind, I'm happy to fix them if you report those bugs to me).

Reply:
Your revised code was fixed, and pushed. Made the adjustment to outputMaybe, to output via the output code, rather than the prototype, which bypassed each modes output. adjusted the endwarning code to blink red/white as otherwise, if one had a red track, no warning would happen. Same if it was a white track.

Reply:
Thank you. It seems you forgot to push however.

Reply:
sorry, will do that now. Got distracted.

Reply:
Pushed